tensorboard/program.py

```diff
@@ -266,61 +265,101 @@ def __init__(self, msg):
+  # ThreadedWSGIServer handles this in werkzeug 0.12+ but we allow 0.11.x.
+  daemon_threads = True
 
   def __init__(self, wsgi_app, flags):
+    self._flags = flags
     host = flags.host
-    port = flags.port
+    self._auto_wildcard = False
+    if not host:
+      # Without an explicit host, we default to serving on all interfaces,
+      # and will attempt to serve both IPv4 and IPv6 traffic through one socket.
+      host = self._get_wildcard_address(flags.port)
+      self._auto_wildcard = True
     try:
-      if host:
-        # The user gave us an explicit host
-        server = serving.make_server(host, port, wsgi_app, threaded=True)
-        if ':' in host and not host.startswith('['):
-          # Display IPv6 addresses as [::1]:80 rather than ::1:80
-          final_host = '[{}]'.format(host)
+      super(WerkzeugServer, self).__init__(host, flags.port, wsgi_app)
+    except socket.error as e:
+      if hasattr(errno, 'EACCES') and e.errno == errno.EACCES:
+        msg = ('TensorBoard must be run as superuser to bind to port %d'
+               % flags.port)
+      elif hasattr(errno, 'EADDRINUSE') and e.errno == errno.EADDRINUSE:
+        if flags.port == 0:
+          msg = 'TensorBoard unable to find any open port'
         else:
-          final_host = host
+          msg = ('TensorBoard could not bind to port %d, it was already in use'
+                 % flags.port)
       else:
-        # We've promised to bind to all interfaces on this host. However, we're
-        # not sure whether that means IPv4 or IPv6 interfaces.
-        try:
-          # First try passing in a blank host (meaning all interfaces). This,
-          # unfortunately, defaults to IPv4 even if no IPv4 interface is available
-          # (yielding a socket.error).
-          server = serving.make_server(host, port, wsgi_app, threaded=True)
-        except socket.error:
-          # If a blank host didn't work, we explicitly request IPv6 interfaces.
-          server = serving.make_server('::', port, wsgi_app, threaded=True)
-        final_host = socket.gethostname()
-      server.daemon_threads = True
-    except socket.error:
-      if port == 0:
-        msg = 'TensorBoard unable to find any open port'
-      else:
-        msg = (
-            'TensorBoard attempted to bind to port %d, but it was already in use'
-            % port)
+        raise
       raise TensorBoardServerException(msg)
-    server.handle_error = _handle_error
-    final_port = server.socket.getsockname()[1]
-    self._server = server
-    self._url = 'http://%s:%d%s' % (final_host, final_port, flags.path_prefix)
 
-  def serve_forever(self):
-    self._server.serve_forever()
+  def _get_wildcard_address(self, port):
+    """Returns a wildcard address for the port in question.
+
+    This will attempt to follow the best practice of calling getaddrinfo() with
+    a null host and AI_PASSIVE to request a server-side socket wildcard address.
```

Comment:
non-AI: where can I read more about `AI_PASSIVE`? Python doc on socket is weird... :(